File: python/demo20b_semaphore.py
'''
SEMAPHORES
Version B: Tires and chassis
'''

import time
import threading



sem_tire = threading.Semaphore(4)
sem_chassis = threading.Semaphore(0)



def make_tire():
    for _ in range(8):
        sem_tire.acquire()

        print('Make 1 tire')
        time.sleep(1)

        sem_chassis.release()



def make_chassis():
    for _ in range(4):
        sem_chassis.acquire()
        sem_chassis.acquire()
        sem_chassis.acquire()
        sem_chassis.acquire()

        print('Make 1 chassis')
        time.sleep(3)

        sem_tire.release()
        sem_tire.release()
        sem_tire.release()
        sem_tire.release()



threading.Thread(target=make_tire).start()
threading.Thread(target=make_tire).start()
threading.Thread(target=make_chassis).start()

File: csharp/demo/demo20b-semaphore.cs
/*
 * SEMAPHORES
 * Version B: Tires and chassis
 */
using System;
using System.Threading;



class Demo20B : IRunnable
{
    public void run()
    {
        var semTire = new Semaphore(4, int.MaxValue);
        var semChassis = new Semaphore(0, int.MaxValue);


        ThreadStart makeTire = () =>
        {
            for (int i = 0; i < 8; ++i)
            {
                semTire.WaitOne();

                Console.WriteLine("Make 1 tire");
                Thread.Sleep(1000);

                semChassis.Release();
            }
        };


        ThreadStart makeChassis = () =>
        {
            for (int i = 0; i < 4; ++i)
            {
                semChassis.WaitOne();
                semChassis.WaitOne();
                semChassis.WaitOne();
                semChassis.WaitOne();

                Console.WriteLine("Make 1 chassis");
                Thread.Sleep(3000);

                semTire.Release();
                semTire.Release();
                semTire.Release();
                semTire.Release();
            }
        };


        new Thread(makeTire).Start();
        new Thread(makeTire).Start();
        new Thread(makeChassis).Start();
    }
}
